File: 프로그래머스/3/43162. 네트워크/네트워크.py
def solution(n, computers):
    answer = 0
    
    parent_set = set()
    
    rank = [0 for _ in range(n)]
    parent = [i for i in range(n)]
    
    for i in range(n):
        for j in range(i + 1, n):
            
            if computers[i][j] == 0:
                continue
            
            union(i, j, parent, rank)
    
    for i in range(n):
        parent_set.add(find(i, parent))
    
    
    return len(parent_set)

def find(node: int, parent: list):
    if node != parent[node]:
        parent[node] = find(parent[node], parent)
    return parent[node]

def union(node1: int, node2: int, parent: list, rank: list):
    root1: int = find(node1, parent)
    root2: int = find(node2, parent)
    
    if rank[root1] > rank[root2]:
        parent[root2] = root1
    else:
        parent[root1] = root2
        if rank[root1] == rank[root2]:
            rank[root2] += 1
    
    

File: 프로그래머스/3/43162. 네트워크/네트워크.cpp
#include <string>
#include <vector>
#include <iostream>
#include <unordered_set>

using namespace std;

int find_node(int node, vector<int>& parent){
    if(node != parent[node])
        parent[node] = find_node(parent[node], parent);
    return parent[node];
}


void union_node(int node1, int node2, vector<int>& parent, vector<int>& rank){
    int root1 = find_node(node1, parent);
    int root2 = find_node(node2, parent);
    
    if (rank[root1] > rank[root2]){
        parent[root2] = root1;
    }
    else{
        parent[root1] = root2;
        if(rank[root1] == rank[root2]){
            rank[root1] ++;
        }
    }
}

int solution(int n, vector<vector<int>> computers) {
    
    vector<int> rank(n, 0);
    vector<int> parent(n, 0);
    
    for(int i = 0; i < n; i++)
        parent[i] = i;
    
    unordered_set<int> connected;
    
    for(int i = 0; i < n; i++){
        for(int j = i + 1; j < n; j++){
            if(computers[i][j] == 0)
                continue;
            
            if(find_node(i, parent) != find_node(j, parent)){
                union_node(i, j, parent, rank);
            }
        }
    }
    
    for(int i = 0; i < n; i++){
        connected.insert(find_node(i, parent));
    }
    
    for (int val : connected) {
        cout << val << " ";
    }
    cout << endl;
    
    return connected.size();
}